### PR TITLE
Runme notebooks launcher

### PR DESCRIPTION
**Pull request by degrammer**
First iteration for having a markdown filter tree as part of runme extension.
TODO:
- [ ] Make folder exclusion configurable (e.g exclude node_modules folder)
- [x] Open markdown files as runme notebooks
- [ ] Make tree navigation more user friendly, reflecting the original folders structure from the workspace
- [ ] Allow right click over specific folders to filter markdown files

Fixes [https://github.com/stateful/vscode-runme/issues/120](https://github.com/stateful/vscode-runme/issues/120)

https://user-images.githubusercontent.com/4001529/212779676-b732bca5-bcf5-412f-83eb-df948edb544e.mp4

### CODE REVIEW

**Comment by sourishkrout** [comment]
Looks like a great start 👍

**Comment by sourishkrout** [comment]
> * [ ]  Make folder exclusion configurable (e.g exclude node_modules folder)
It'd be great if that was a simple toggle like the timeline panel offers
image

> * [ ]  Open markdown files as runme notebooks
Yep, that's a must have

> * [ ]  Make tree navigation more user friendly, reflecting the original folders structure from the workspace
Let's definitely experiment with that. Currently the UX is a little rocky if entries aren't uncollapsed by default.

> * [ ]  Allow right click over specific folders to filter markdown files
I'd put this in the nice to have category.

Outside of the todos, I'd like to do some editing on panel name etc closer to the finish line.

File: src/extension/provider/launcher.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    return Promise.resolve(folderMarkdownItems)
+  }
+
+  refresh(): void {
+    this._onDidChangeTreeData.fire()
+  }
+
+  getRunmeFilesFromWorkspace(onComplete: (value: RunmeFile[]) => void): any {
+    glob('**/*.md', { cwd: this.workspaceRoot, ignore: 'node_modules/**', absolute: true }, (err, files) => {
```
Comment (sourishkrout) [inline src/extension/provider/launcher.ts:90]: @degrammer you do want to use the VS Code API's way to glob/traverse the FS. The reason is that vscode can run against virtual filesystems when attached to a SSH host or directly attached to a Github Repository. It's really only guaranteed that local FS is available when VS Code is run as Desktop app. It's something to get used to building inside of VS Code.

image
Can't deep link to the API hence screenshot: https://code.visualstudio.com/api/references/vscode-api#workspace

File: src/extension/provider/launcher.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
+import { join, basename, dirname, resolve } from 'node:path'
+
+import * as vscode from 'vscode'
```
Comment (christian-bromann) [inline src/extension/provider/launcher.ts:3]:
```suggestion
import vscode from 'vscode'
```

File: tests/extension/provider/launcher.test.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { vi, describe, it, expect } from 'vitest';
```
Comment (sourishkrout) [inline tests/extension/provider/launcher.test.ts:1]: Nit: Looks like this file need a eslint fix-all run. Based on the current rules I won't pass linting.

File: package.json [hunk 4]
```diff
@@ -263,6 +263,20 @@
           }
         }
       }
+    ],
+    "views": {
+      "explorer": [
+        {
+          "id": "runme.launcher",
+          "name": "Runme Notebooks"
```
Comment (sourishkrout) [inline package.json:271]: I believe there's an attribute that controls default visibility (collapsed/uncollapsed). We probably want to decide how front and center we want to be?

File: package.json [hunk 5]
```diff
@@ -275,6 +289,7 @@
     "postinstall": "npm run download:wasm",
     "test": "run-s test:*",
     "test:lint": "eslint src --ext ts",
+    "test:lint:fix": "eslint src --ext ts --fix",
```
Comment (sourishkrout) [inline package.json:292]: This won't catch the `tests` directory :-D

**Review by sourishkrout — CHANGES_REQUESTED** [review]
Looks fantastic 👏 ! My and Christian's comments are just nits. Should be easily fixed and ready for merge 👍  after (no additional review required).

In future enhancements (just writing down here so we don't forget):
- Consider adding the treeview highlighting the currently active notebook file
- Action buttons to quickly collapse/uncollapse all entries analogous to the file explorer.

Let's port these items plus the open check boxes above into a new issue for future feature consideration.

File: package.json (hunk 4 above)
Comment (degrammer) [reply package.json]: Added a default collapsed visibility

File: package.json (hunk 5 above)
Comment (degrammer) [reply package.json]: Right, fixed :)

File: tests/extension/provider/launcher.test.ts (hunk 3 above)
Comment (degrammer) [reply tests/extension/provider/launcher.test.ts]: Fixed

File: src/extension/provider/launcher.ts (hunk 2 above)
Comment (degrammer) [reply src/extension/provider/launcher.ts]: Fixed, thanks

**Pull request merged**